src/v/cluster/cloud_storage_size_reducer.h

```diff
@@ -0,0 +1,196 @@
+      , _retries_allowed{retries_allowed} {}
+
+    ss::future<std::optional<uint64_t>> reduce() {
+        try {
+            co_return co_await do_reduce();
+        } catch (const topic_table_partition_generator_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const cloud_storage_size_reducer_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const std::exception& e) {
+            vlog(clusterlog.warn, "Unexpected exception thrown: {}", e.what());
+            throw;
+        }
+
+        if (++_retries_attempted <= _retries_allowed) {
+            co_return co_await reduce();
+        } else {
+            co_return std::nullopt;
+        }
+    }
+
+private:
+    ss::future<uint64_t> do_reduce() {
+        topic_table_partition_generator partition_gen{
+          _topic_table, _batch_size};
+
+        uint64_t total_cloud_storage_bytes{0};
+        auto batch = co_await partition_gen.next_batch();
+        while (batch) {
+            requests_t request_futures = map_batch(*batch);
+            auto results = co_await ss::when_all_succeed(
+              request_futures.begin(), request_futures.end());
+
+            for (auto& res : results) {
+                if (res.has_error()) {
+                    throw cloud_storage_size_reducer_exception(
+                      "RPC for cloud storage usage failed");
+                }
+
+                if (res.value().missing_partitions.size() > 0) {
+                    throw cloud_storage_size_reducer_exception(fmt::format(
+                      "Cloud storage usage for {} partitions was not found",
+                      res.value().missing_partitions.size()));
+                }
+
+                total_cloud_storage_bytes += res.value().total_size_bytes;
+            }
+
+            batch = co_await partition_gen.next_batch();
+        }
+
+        co_return total_cloud_storage_bytes;
+    }
+
+    ss::future<result<cloud_storage_usage_reply>> send_query(
+      const model::broker& destination, cloud_storage_usage_request req) {
+        return with_client<controller_client_protocol>(
+          _self.id(),
+          _conn_cache,
+          destination.id(),
+          destination.rpc_address(),
+          _rpc_tls_config,
+          _timeout,
+          [this, req = std::move(req)](controller_client_protocol c) mutable {
+              return c
+                .cloud_storage_usage(std::move(req), rpc::client_opts(_timeout))
+                .then(&rpc::get_ctx_data<cloud_storage_usage_reply>);
```

Comment:
Could there be exceptions thrown here? Might be able to check that in the above method and return retry-able error if so

Reply:
If `send_query` results in a an exceptional future, `do_reduce` throws a retry-able exception. If an actual exception is thrown, it will bubble up from `reduce` via the `Unexpected exception` path. I think this is fine as `send_query` shouldn't really throw, and, if it does, it means something is intrinsically wrong and we should abort (e.g. bad allocs).